lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1239 @@
+        (:math:`f_h`).
+    nonlinearity : callable or None
+        Nonlinearity to apply when computing new state (:math:`\sigma`). If
+        None is provided, no nonlinearity will be applied.
+    hid_init : callable, np.ndarray, theano.shared or TensorVariable
+        Passing in a TensorVariable allows the user to specify
+        the value of `hid_init` (:math:`h_0`). In this mode, `learn_init` is
+        ignored.
+    backwards : bool
+        If True, process the sequence backwards and then reverse the
+        output again such that the output from the layer is always
+        from :math:`x_1` to :math:`x_n`.
+    learn_init : bool
+        If True, initial hidden values are learned. If `hid_init` is a
+        TensorVariable then the TensorVariable is used and
+        `learn_init` is ignored.
+    gradient_steps : int
+        Number of timesteps to include in the backpropagated gradient.
+        If -1, backpropagate through the entire sequence.
+    grad_clipping : False or float
+        If a float is provided, the gradient messages are clipped during the
+        backward pass.  If False, the gradients will not be clipped.  See [1]_
+        (p. 6) for further explanation.
+    unroll_scan : bool
+        If True the recursion is unrolled instead of using scan. For some
+        graphs this gives a significant speed up but it might also consume
+        more memory. When `unroll_scan` is true then the `gradient_steps`
+        setting is ignored. The input sequence length cannot be specified as
+        None when `unroll_scan` is True.
+    precompute_input : bool
+        If True, precompute input_to_hid before iterating through
+        the sequence. This can result in a speedup at the expense of
+        an increase in memory usage.
+
+    References
+    ----------
+    .. [1] Graves, Alex: "Generating sequences with recurrent neural networks."
+           arXiv preprint arXiv:1308.0850 (2013).
+    """
+    def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
+                 nonlinearity=nonlinearities.rectify,
+                 hid_init=init.Constant(0.),
+                 backwards=False,
+                 learn_init=False,
+                 gradient_steps=-1,
+                 grad_clipping=False,
+                 unroll_scan=False,
+                 precompute_input=True,
+                 **kwargs):
+
+        super(CustomRecurrentLayer, self).__init__(incoming, **kwargs)
+
+        self.input_to_hidden = input_to_hidden
+        self.hidden_to_hidden = hidden_to_hidden
+        self.learn_init = learn_init
+        self.backwards = backwards
+        self.gradient_steps = gradient_steps
+        self.grad_clipping = grad_clipping
+        self.unroll_scan = unroll_scan
+        self.precompute_input = precompute_input
+
+        if unroll_scan and gradient_steps != -1:
+            raise ValueError(
+                "Gradient steps must be -1 when unroll_scan is true.")
```

Comment:
Ah, this is not in line with the documentation. It's a lot better this way, though. Can you ignore part of my suggestion above and change the documentation to say: "When unroll_scan is True, backpropagation always includes the full sequence, so gradient_steps must be set to -1 and the input sequence length must be known at compile time (i.e., cannot be given as None)."